### PR TITLE
[SPARK-15474][SQL] Write and read back non-emtpy schema with empty dataframe

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Previously, ORC file format cannot write a correct schema in case of empty dataframe. Instead, it creates an empty ORC file with **empty** schema, `struct<>`. So, Spark users cannot write and read back ORC files with non-empty schema and no rows. This PR uses new Apache ORC 1.4.1 to create an empty ORC file with a correct schema. Also, this PR uses ORC 1.4.1 to infer schema always.

**BEFORE**
```scala
scala> val emptyDf = Seq((true, 1, "str")).toDF("a", "b", "c").limit(0)
scala> emptyDf.write.format("orc").mode("overwrite").save("/tmp/empty")
scala> spark.read.format("orc").load("/tmp/empty").printSchema
org.apache.spark.sql.AnalysisException: Unable to infer schema for ORC. It must be specified manually.;
```

**AFTER**
```scala
scala> spark.read.format("orc").load("/tmp/empty").printSchema
root
 |-- a: boolean (nullable = true)
 |-- b: integer (nullable = true)
 |-- c: string (nullable = true)
```

## How was this patch tested?

Pass the Jenkins with newly added test cases.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala [hunk 1]
```diff
@@ -39,4 +45,33 @@ private[sql] object OrcFileFormat {
...
+      case _: IOException => None
+    }
+  }
+
+  def readSchema(sparkSession: SparkSession, files: Seq[FileStatus]): Option[StructType] = {
+    val conf = sparkSession.sparkContext.hadoopConfiguration
+    val fs = FileSystem.get(conf)
+    val options = OrcFile.readerOptions(conf).filesystem(fs)
+    files.map(_.getPath).flatMap(readSchema(_, options))
+      .headOption.map { schema =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala]: Seems that you just take the first available schema. Looks like we don't need to read other files when we found the first available schema.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 2]
```diff
@@ -252,6 +253,13 @@ private[orc] class OrcOutputWriter(
   override def close(): Unit = {
     if (recordWriterInstantiated) {
       recordWriter.close(Reporter.NULL)
+    } else {
+      // SPARK-15474 Write empty orc file with correct schema
+      val conf = context.getConfiguration()
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Looks like the behavior to skip creating an empty file if no rows are written is deliberate. Is there any impact to current behavior?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 3]
```diff
@@ -252,6 +253,13 @@ private[orc] class OrcOutputWriter(
   override def close(): Unit = {
     if (recordWriterInstantiated) {
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Btw, according to the existing comment, seems that we can simply remove `recordWriterInstantiated` to allow empty file created.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 4]
```diff
@@ -252,6 +253,13 @@ private[orc] class OrcOutputWriter(
...
       recordWriter.close(Reporter.NULL)
+    } else {
+      // SPARK-15474 Write empty orc file with correct schema
+      val conf = context.getConfiguration()
+      val writer = org.apache.orc.OrcFile.createWriter(
+        new Path(path), org.apache.orc.mapred.OrcOutputFormat.buildOptions(conf))
+      new org.apache.orc.mapreduce.OrcMapreduceRecordWriter(writer)
+      writer.close()
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: So, if i understood correctly it will write out by `org.apache.orc.mapreduce.OrcMapreduceRecordWriter` when output is empty but, write out by `org.apache.hadoop.hive.ql.io.orc.OrcRecordWriter` when output is non-empty? I thought we should use the same writer for both paths if possible and this one looks rather a band-aid fix. It won't block this PR but I wonder if this is the only way we could do for now.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 5]
```diff
@@ -73,6 +70,10 @@ class OrcFileFormat extends FileFormat with DataSourceRegister with Serializable
 
     val configuration = job.getConfiguration
 
+    configuration.set(
+      MAPRED_OUTPUT_SCHEMA.getAttribute,
+      org.apache.spark.sql.execution.datasources.orc.OrcFileFormat.getSchemaString(dataSchema))
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Do we always need to set this?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 6]
```diff
@@ -2127,4 +2127,18 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
       }
     }
   }
+
+  Seq("orc", "parquet").foreach { format =>
+    test(s"SPARK-15474 Write and read back non-emtpy schema with empty dataframe - $format") {
+      withTempDir { dir =>
+        val path = dir.getCanonicalPath
+        val emptyDf = Seq((true, 1, "str")).toDF.limit(0)
+        emptyDf.write.format(format).mode("overwrite").save(path)
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Hm, why is `withTempPath { path =>` not used without `overwrite` instead?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 7]
```diff
@@ -58,10 +58,7 @@ class OrcFileFormat extends FileFormat with DataSourceRegister with Serializable
       sparkSession: SparkSession,
       options: Map[String, String],
       files: Seq[FileStatus]): Option[StructType] = {
-    OrcFileOperator.readSchema(
-      files.map(_.getPath.toString),
-      Some(sparkSession.sessionState.newHadoopConf())
-    )
+    org.apache.spark.sql.execution.datasources.orc.OrcFileFormat.readSchema(sparkSession, files)
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: I am not sure of this one too. This looks a complete rewrite of `org.apache.spark.sql.hive.orc.OrcFileOperator.readSchema`.. Is this change required to fix this issue?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @viirya and @HyukjinKwon .
You know that I tried to introduce new OrcFileFormat in `sql/core` before. But, it is too big for review. According to @cloud-fan 's advice, I'm trying to upgrade the existing one by one in a piece.

So far,
- We introduced new ORC 1.4.1 dependency
- Introduce new Spark SQL ORC parameters and replace Hive constant with new ORC parameters.

This is the actual first PR to use read and write using ORC 1.4.1 library.
- It reads ORC file only for inferencing schema.
- It writes only empty ORC file.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala]: Yes. This is based on the existing [OrcFileOperator.readSchema](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileOperator.scala#L74-L82).

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Previously, only ORC does. So, it creates more issues like SPARK-22258 (#19477) and SPARK-21762. This is consistent with the other data sources like Parquet.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: In this PR, I'm focusing on emtpy file. We will replace the whole writer and reader with ORC 1.4.1 eventually. The newly added test case in this PR will make us to transit safely.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 7 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Yes. It's intentional. OrcFileOperator will be replaced later completely. I made this PR as small as possible for review.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Yes. This is the correct schema to be written.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Yep. That's correct understanding. This PR intentionally focuses only on handling empty files and inferring schema. This will help us transit safely from old Hive ORC to new Apache ORC 1.4.1.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Thanks. No problem. I'll use `withTempPath`.

**Comment by dongjoon-hyun** [comment]
I updated the PR.
Could you review this PR again, @viirya , @HyukjinKwon , @gatorsmile , @cloud-fan ?

**Comment by gatorsmile** [comment]
What is the backward compatibility of ORC  1.4.1? Can we create multiple ORC files created by the previous versions and ensure they are not broken?

**Comment by cloud-fan** [comment]
I checked with how we introduce the new parquet reader before, and maybe we can follow it: https://github.com/apache/spark/pull/4308

Basically we leave the old orc data source as it is, and implement a new orc 1.4.1 data source in sql core module. Then we have an internal config to switch the implementation(by default true), and remove the old implementation after one or two releases.

**Comment by cloud-fan** [comment]
> What is the backward compatibility of ORC 1.4.1? Can we create multiple ORC files created by the previous versions and ensure they are not broken?

That a good point, and I think it's better to have these tests in the orc project. If they don't have, then we can take over and add these tests.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @gatorsmile and @cloud-fan . Especially, @cloud-fan 's opinion is my original approach in #17980 and #18953 (before Aug 16). I cannot agree any more.

> Basically we leave the old orc data source as it is, and implement a new orc 1.4.1 data source in sql core module. Then we have an internal config to switch the implementation(by default prefer the new implementation), and remove the old implementation after one or two releases.

BTW, I'm wondering what is changed after you commented [the following](https://github.com/apache/spark/pull/18953#issuecomment-322827590) on that PR on 16th Aug.

> Are the ORC APIs changed a lot in 1.4? I was expecting a small patch to upgrade the current ORC data source, without moving it to sql/core.

**Comment by dongjoon-hyun** [comment]
@gatorsmile and @cloud-fan . 
For ORC compatibility, I checked the ORC code, but it's not clearly tested.
I'll try to add some suite as a separate issue.

**Comment by dongjoon-hyun** [comment]
To be clear, for [ORC File Versions](https://github.com/apache/orc/blob/master/proto/orc_proto.proto#L240-L248), there exists some ORC test case against version 0.11, but it's not our scope because Spark (and Hive 1.2) uses `0.12 with HIVE_8732`.

There are 6 versions with 0.12.

0 = original
**1 = HIVE-8732 fixed (fixed stripe/file maximum statistics & string statistics use utf8 for min/max)**
2 = HIVE-4243 fixed (use real column names from Hive tables)
3 = HIVE-12055 fixed (vectorized writer implementation)
4 = HIVE-13083 fixed (decimals write present stream correctly)
5 = ORC-101 fixed (bloom filters use utf8 consistently)
6 = ORC-135 fixed (timestamp statistics use utc)

**Comment by cloud-fan** [comment]
Sorry I miss-understood the problem at the beginning. I thought the new orc version just changes the existing APIs a little, but it turns out the new orc version has a new set of read/write APIs.

**Comment by dongjoon-hyun** [comment]
I see. Then, can we continue on #17980 ?

**Comment by cloud-fan** [comment]
yes please

**Comment by dongjoon-hyun** [comment]
This is resolved in https://github.com/apache/spark/pull/19651 .
